salt/pillar/hg_pillar.py

```diff
@@ -0,0 +1,139 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2014 Floris Bruynooghe <flub@devork.be>
+
+'''Use remote Mercurial repository as a Pillar source
+
+The module depends on the ``hglib`` python module being available.
+This is the same requirement as for hgfs_ so should not pose any extra
+hurdles.
+
+This external Pillar source can be configued in the master config file as such:
+
+.. code-block:: yaml
+
+   ext_pillar:
+     - hg: ssh://hg@example.co/user/repo
+'''
+
+import copy
+import hashlib
+import logging
+import os
+
+import salt.pillar
+
+try:
+    import hglib
+except ImportError:
+    hglib = None
+
+
+__virtualname__ = 'hg'
+log = logging.getLogger(__name__)
+
+
+# The default option values
+__opts__ = {}
+
+
+def __virtual__():
+    '''
+    Only load if hglib is available
+    '''
+    ext_pillar_sources = [x for x in __opts__.get('ext_pillar', [])]
+    if not any(['hg' in x for x in ext_pillar_sources]):
+        return False
+    if not hglib:
+        log.error('hglib not present')
+        return False
+    return __virtualname__
+
+
+def __init__(__opts__):
+    '''
+    Initialise
+
+    This is called every time a minion calls this external pillar.
+    '''
+
+
+def ext_pillar(minion_id, pillar, repo, branch='default', root=None):
+    '''
+    Extract pillar from an hg repository
+    '''
+    with Repo(repo) as repo:
+        repo.update(branch)
+    envname = 'base' if branch == 'default' else branch
+    if root:
+        path = os.path.normpath(os.path.join(repo.working_dir, root))
+    else:
+        path = repo.working_dir
+
+    opts = copy.deepcopy(__opts__)
+    opts['pillar_roots'][envname] = [path]
+    pil = salt.pillar.Pillar(opts, __grains__, minion_id, envname)
+    return pil.compile_pillar(ext=False)
+
+
+def update(repo_uri):
+    '''
+    Execute an hg pull on all the repos
+    '''
+    with Repo(repo_uri) as repo:
+        repo.pull()
+
+
+def envs():
+    '''
+    Return a list of branches that can be used as environments
+    '''
+
+
+def purge_cache():
+    '''
+    Purge the hg_pillar cache
+    '''
+
```

Comment:
I believe they are for future implementation, but should be removed as they are not called in the hg_pillar.py. I am testing if things are working correctly without them